feat(parser): add a shapefile parser

zarov opened the pull request:
The parser can take into account four files: shp, shx, dbf and prj. It
uses https://www.npmjs.com/package/shpjs

An example and some data has been added.

This PR should not be merged before #1033

zarov committed:
feat(parser): add a simple Shapefile parser

The parser can take into account four files: shp, shx, dbf and prj. It
uses https://www.npmjs.com/package/shpjs

An example and some data has been added.

zarov commented:
This PR can be reviewed now that #1033 has been merged.

gchoqueux reviewed (commented):
could you move data files in itowns sample data?

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,79 @@
+<html>
```
add in index.html + add image

zarov replied:
Done

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,79 @@
+
+            // Load all the necessary files for a shapefile, parse them and
+            // display them.
+            Promise.all([
```
could you add `Fetcher.Shape(baseUrl, extentionfiles = {shp: 'shp',.... }, networkOptions)`?

zarov replied:
So I pushed a quick draft, let me know what you think

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,79 @@
+                }, {
+                    buildExtent: true,
+                    crsIn: 'EPSG:4326',
+                    crsOut: view.tileLayer.extent.crs(),
```
could you remove `crsOut`?

zarov replied:
I can't ? It is needed by `GeoJsonParser`.

gchoqueux commented on src/Parser/ShapefileParser.js:
```diff
@@ -0,0 +1,78 @@
+            ]).then(shp.combine);
+        }
+
+        options.crsIn = data.prj ? proj4(data.prj).oProj.datumName : undefined;
```
why define `crsIn`  in example?

zarov replied:
Removed it !

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,70 @@
+                text: ['prj'],
+            }).then(function _(res) {
+                return itowns.ShapefileParser.parse(res, {
+                    buildExtent: true,
```
Sorry, effectively, `crsout` is necessary but  i think it's not user  friendly to write `view.extent.crs()`. 

In future, may be we'll add `view.avaibleProjectionColor()`.

zarov replied:
Yes sorry I added it back. Yeah I agree, we may find something simpler.

zarov commented:
I completed the documentation and rebased everything in 2 commits

zarov committed:
feat(fetcher): add a multiple method to load a bunch of files

The new multiple method can be used to load a bunch of files having the
same base url, like the shapefile system for example.

gchoqueux commented:
I have some comments:

- The example doesn't load, it needs to move to refresh globe. could you fix this issue?
- I remarked the points circle are deformed in the direction of latitude because the texture have a size squared. could you open a issue? a solution to fix it, is to use PM projection.
- The drawing of points is very slow because there is not test on extent. I have start a optimisation could add it?  
[0001-optimization-draw-point.patch](https://github.com/iTowns/itowns/files/2956137/0001-optimization-draw-point.txt)

zarov commented:
- This has been adressed in #1073 
- I can't find how to apply the correct projection, sorry. I'm opening an issue on this.
- Done

zarov committed:
perf(converter): add extent checks in Feature2Texture

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,71 @@
+            });
+            // Add two elevation layers.
+            // These will deform iTowns globe geometry to represent terrain elevation.
+            function addElevationLayerFromConfig(config) {
```
`addElevationLayerFromConfig` is useless.

zarov replied:
removed

gchoqueux commented on src/Converter/Feature2Texture.js:
```diff
@@ -65,27 +66,37 @@ function drawPoint(ctx, x, y, origin, scale, style = {}) {
 function drawFeature(ctx, feature, origin, scale, extent, style = {}) {
     let gStyle = style;
+    // Determinate with radius point
+    const px = 0.01;
```
`px` needs to be determinate with radius, because if radius increases then the condition is wrong.

zarov replied:
Damn I copied/pasted it without noticing the comment, my bad ! Fixing it.

zarov replied:
So this seems complicated to do: it depends a lot on the current projection/view, and I don't think this will simply things. What about allowing only a maximum radius size ? I don't see a case where a point is larger than 30 or 50px, and unless someone displays this point on a globe with a zoom at 0, there will be no artefact. What do you say ?

zarov replied:
Nervemind, it is done simply in fact, my bad

gchoqueux commented on examples/shapefile.html:
```diff
@@ -0,0 +1,71 @@
+                    buildExtent: true,
+                    crsOut: view.tileLayer.extent.crs(),
+                });
+            }).then(function _(geojson) {
```
As the parser returns a `Feature`, could you rename,  `geojson`, in `feature`?

zarov replied:
Done

gchoqueux commented:
> The example doesn't load, it needs to move to refresh globe. could you fix this issue?

fixed with #1068

gchoqueux commented on src/Converter/Feature2Texture.js:
```diff
@@ -65,27 +66,45 @@ function drawPoint(ctx, x, y, origin, scale, style = {}) {
+
+        // cross multiplication to know in the extent system the real size of
+        // the point
+        px = gStyle.radius * (extentDim.x / 256);
```
you could use `ctx.canvas.width` instead of `256`
